packages/scm/src/browser/scm-tree-model.ts

```diff
@@ -207,19 +211,19 @@ export class ScmTreeModel extends TreeModelImpl {
                 if (folderEnd - folderStart < nestingThreshold) {
                     // Inline these (i.e. do not create another level in the tree)
                     for (let i = folderStart; i < folderEnd; i++) {
-                        result.push(this.toFileChangeNode(resources[i].resource, parent));
+                        result.push(this.toFileChangeNode(sortedResources[i].resource, parent));
                     }
                 } else {
                     const firstFileParts = firstFileChange.pathParts;
-                    const lastFileParts = resources[folderEnd - 1].pathParts;
+                    const lastFileParts = sortedResources[folderEnd - 1].pathParts;
                     // Multiple files with first folder.
                     // See if more folder levels match and include those if so.
                     let thisLevel = level + 1;
                     while (thisLevel < firstFileParts.length - 1 && thisLevel < lastFileParts.length - 1 && firstFileParts[thisLevel] === lastFileParts[thisLevel]) {
                         thisLevel++;
                     }
                     const nodeRelativePath = firstFileParts.slice(level, thisLevel).join('/');
-                    result.push(this.toFileChangeFolderNode(resources, folderStart, folderEnd, thisLevel, nodeRelativePath, parent));
+                    result.push(this.toFileChangeFolderNode(sortedResources, folderStart, folderEnd, thisLevel, nodeRelativePath, parent));
                 }
                 folderStart = folderEnd;
             }
```

Comment:
I think we can get rid of `result.sort(...)` if we keep separate result arrays for file nodes and folder nodes and concatenate them at the end. With your change, the list of resources is already sorted and as far as I can see, the sorting will be correct under the algorithm.